Feature/expansive sdf

tcmoore3 opened the pull request:
<!-- Please confirm that your work is based on the correct branch. -->
<!-- Bug fixes should be based on *trunk-patch*. -->
<!-- Backwards compatible new features should be based on *trunk-minor*. -->
<!-- Incompatible API changes should be based on *trunk-major*. -->

## Description

Adds expansive volume perturbations to `ComputeSDF`. 

## Motivation and context

This is a follow-up to #1356 that enables pressure computation for systems of concave particles. 

## How has this been tested?

I tested this code in a separate branch before adding it to this branch by running simulations of a 2D sphere union in and confirming pressure–density consistency between NVT and NPT simulations. I will re-run that test using this branch and post the results when complete.

## Change log

<!-- Propose a change log entry. -->
```
- Enable computation of pressure in HPMC simulations of nonconvex particles
```

## Checklist:

- [x] I have reviewed the [**Contributor Guidelines**](https://github.com/glotzerlab/hoomd-blue/blob/trunk-minor/CONTRIBUTING.md).
- [x] I agree with the terms of the [**HOOMD-blue Contributor Agreement**](https://github.com/glotzerlab/hoomd-blue/blob/trunk-minor/ContributorAgreement.md).
- [x] My name is on the list of contributors (`sphinx-doc/credits.rst`) in the pull request source branch.
- [ ] Post validation test results
- [x] Update python documentation
- [x] Figure out what unit tests are needed and implement them

tcmoore3 committed:
Compiles but is untested

tcmoore3 committed:
Compiles but is untested

tcmoore3 committed:
Implement expansion moves for SDF

The SDF method that calculates the pressure in systems with
discontinuous potentials (e.g., hard particles) uses volume
perturbations. As originally implemented for convex hard particles, only
negative volume perturbations are implemnted since they are the only
volume prturbations that contribute to the pressure in those systems.
However, some systems, e.g., systems with discontinuous pair potentials
or convex hard particles, will have meaningful contributions to the
pressure from epansive volume perturbations.

This commit implements the expansion moves in addition to the
compression moves. There is a lot of code duplication because I was
first testing whether or not we can get reasonable answers before making
the code more elegant. So now we keep track of two scale distribution
functions: one for expansive and one for compressive volume
perturbations. The pressure that is returned is a weighted average of
the pressure calculated from expansion and compression moves, weighted
by the number of entries in each respective histogram. However, there is
nothing from the theory that says this should be a weighted average, and
instead should be an unweighted average; I used the weighted average
because that gave me pressures that were closer to the "ground truth"
pressure in my tests. Hence, use this code at your own risk. Further
theoretical developments will work address this inconsistency between
theory and simulation.

tcmoore3 committed:
Correctly compute pressure with expansion moves

The correct way to compute pressure when doing both expansion and
compression moves is to simply sum their contributions to the excess
pressure. This algorithm gives the expected result for hard convex
particles where isotropic expansions cannot result in overlaps, so the
contribution to the (excess) pressure is zero for those moves. I am able
to get pressure/density agreement with constant-volume and
constant-pressure simulations for a simple concave paticle model.

tcmoore3 committed:
Merge remote-tracking branch 'origin/trunk-minor' into feature/enthalpic-sdf

tcmoore3 committed:
Add expansion moves to ComputeSDF

This commit adds expansive volume perturbations to the SDF compute.
These perturbations are necessary for nonconvex hard shapes or patchy
"nonmonotonic" patchy interactions (monotonic is probably not the right
word here, but we have not been able to come up with a better word; see
https://github.com/glotzerlab/hoomd-blue/pull/1356#discussion_r943488535
for some discussion on this topic).

To check if expansion moves are need when the compute is triggered, I
added a requiresExpansiveSDF() method to IntegratorHPMCMono and the
different shape classes. The integrator's method returns `true` if the
method for any of the shapes in the system return `true`. The relevant
member variable of the integrator is set during prepRun(). Whenever the
linear search patch is taken, expansion moves are performed. The linear
search path is always taken if either the integrator's
requiresExpansiveSDF method returns true or if there are patchy
interactions in the system, and therefore systems with patches always do
expansion moves.

tcmoore3 committed:
Merge remote-tracking branch 'origin/trunk-minor' into feature/expansive-sdf

tcmoore3 committed:
Tidy up a bit

joaander commented on hoomd/hpmc/ComputeSDF.h:
```diff
@@ -149,14 +149,17 @@ template<class Shape> class ComputeSDF : public Compute
 
     //! Return an sdf
     virtual pybind11::array_t<double> getSDF();
+    virtual pybind11::array_t<double> getSDF_expansion();
```
Follow the naming conventions outline in https://hoomd-blue.readthedocs.io/en/v3.4.0/style.html

joaander commented on hoomd/hpmc/ComputeSDF.h:
```diff
@@ -171,6 +174,7 @@ template<class Shape> class ComputeSDF : public Compute
     void countHistogramLinearSearch(uint64_t timestep);
 
     //! Determine the s bin of a given particle pair
+    //! Only used for the binary search mode
```
It is not valid to provide two brief descriptions `//!` for the same method.

joaander commented on hoomd/hpmc/ComputeSDF.h:
```diff
@@ -149,14 +149,17 @@ template<class Shape> class ComputeSDF : public Compute
 
-    std::vector<double> m_hist; //!< Raw histogram data
-    std::vector<double> m_sdf;  //!< Computed SDF
+    std::vector<double> m_hist;           //!< Raw histogram data
+    std::vector<double> m_sdf;            //!< Computed SDF
```
Will need to name these and related variables with `_compression` to avoid ambiguity.

joaander commented on hoomd/hpmc/ComputeSDF.h:
```diff
@@ -305,7 +335,7 @@ template<class Shape> Scalar ComputeSDF<Shape>::getMaxInteractionDiameter()
 template<class Shape> void ComputeSDF<Shape>::countHistogram(uint64_t timestep)
     {
-    if (m_mc->getPatchEnergy())
+    if (m_mc->getPatchEnergy() || m_mc->requiresExpansiveSDF())
```
There is no need to introduce a new method in every shape. Use C++ constexpr to determine the expansive flag based on the type of `Shape`. For safety, default to `true` and only set to `false` for known convex `Shape` classes.

joaander commented on hoomd/hpmc/ComputeSDF.h:
```diff
@@ -518,61 +550,123 @@ template<class Shape> void ComputeSDF<Shape>::countHistogramLinearSearch(uint64_
+                            // overlap of particle i with its neighbors
+                            bool do_compression = true;
+                            bool do_expansion = true;
+                            for (size_t bin_to_sample = 0;
```
This would be more readable (and a smaller diff) if split into two separate for loops instead of one for loop that conspicuously goes to the max(i) and manages the different paths with bool flags.

joaander commented on hoomd/hpmc/IntegratorHPMCMono.h:
```diff
@@ -346,6 +346,16 @@ class IntegratorHPMCMono : public IntegratorHPMC
                     break;
                     }
                 }
+            m_requiresExpansiveSDF = false;
```
There are no changes needed to this file or the `Shape*` files. As discussed above, detect based on the type of the `Shape` class.

joaander commented on hoomd/hpmc/compute.py:
```diff
@@ -311,16 +311,32 @@ def sdf(self):
 
+    @log(category='sequence', requires_run=True)
+    def sdf_expansion(self):
+        """(*N_bins*,) `numpy.ndarray` of `float`): :math:`s[k]` - The scale \
+        distribution function :math:`[\\mathrm{probability\\ density}]`.
+
+        The :math:`x` at the center of bin :math:`k` is:
+        :math:`x = k \\cdot \\delta x + \\delta x/2`.
+
+        Attention:
+            In MPI parallel execution, the array is available on rank 0 only.
+            `sdf` is `None` on ranks >= 1.
```
This appears to be copied and pasted from `sdf` without modification. Make the necessary changes to correctly document `sdf_expansion`.

joaander commented on hoomd/hpmc/compute.py:
```diff
@@ -311,16 +311,32 @@ def sdf(self):
         self._cpp_obj.compute(self._simulation.timestep)
```
Rename `sdf` to `sdf_compression`. Create a deprecated `sdf` that aliases it. Document that deprecation in `deprecated.rst`.

joaander commented on hoomd/hpmc/compute.py:
```diff
@@ -311,16 +311,32 @@ def sdf(self):
         self._cpp_obj.compute(self._simulation.timestep)
         return self._cpp_obj.sdf
```
I can't comment further up. In the class level docstring, document the expansion histogram and how `\beta P` relates to both the compression and expansion sdfs.

joaander commented on hoomd/hpmc/compute.py:
```diff
@@ -329,19 +345,37 @@ def betaP(self):  # noqa: N802 - allow function name
+            p0_compression = numpy.polyval(p, 0.0)
+            compression_contribution = rho * p0_compression / (2
+                                                               * box.dimensions)
+        if not numpy.isnan(self.sdf_expansion).all():
```
Why are any of the sdf values `nan` in the first place? They are initialized to 0.

tcmoore3 replied:
That was already in the code, so I can't say for sure why it was added. But you're right, they're initialized to 0 and the only way I see they could become `nan` is if `m_pdata->getNGlobal() * m_dx == 0`.

joaander commented on hoomd/hpmc/compute.py:
```diff
@@ -329,19 +345,37 @@ def betaP(self):  # noqa: N802 - allow function name
+        if not numpy.isnan(self.sdf_expansion).all():
+            # get the values to fit
+            # reverse the sdf so that it starts at 0 and goes negative
+            sdf_fit_expansion = self.sdf_expansion[0:n_fit][::-1]
```
Should this be the natural representation that we provide to users? Or do you expect that users may want to plot `[sdf_expansion] + [sdf_compression]` vs some x that goes from `-xmax` to `+xmax`?

joaander replied:
Here is another thought. Instead of providing `sdf_compression` and `sdf_expansion` to the user, we could provide one `sdf` that covers both. We should consider what is the better API for users.

joaander reviewed (changes requested):
Thanks! I have many comments, questions, and suggestions for improvement.

tcmoore3 commented:
My validation tests indicate that I introduced a bug in between my initial implementation and this pull request. Currently investigating.

tcmoore3 committed:
Fix a bug in linear search SDF calculation

I was getting the wrong pressures because of a bug I introduced when
adding the expansion moves. The bug was that I was not carrying over the
minimum distance to overlap from the i-j1 pair to i-j2. Hence, the SDF I
was calculating was too high at larger scale values than it should have
been, and hence the pressure was being underestimated compared to the
true value. This commit fixes that bug by comparing the current bin
(corresponding to the scale value) to the minimum bin where an overlap
has been found and skipping the overlap calculation if the current bin
is greater than the minimum bin.

tcmoore3 committed:
Fix pressure computation

tcmoore3 committed:
Follow naming convention for getSDF_expansion

tcmoore3 committed:
Only 1 single-line comment per thing to document

tcmoore3 committed:
Add _compression to m_hist and m_sdf

To reduce ambiguity now that we have members named m_*_expansion

tcmoore3 committed:
Use Shape template parameter to determine SDF moves

tcmoore3 committed:
Merge branch 'trunk-minor' into feature/expansive-sdf

tcmoore3 committed:
Do expansions and compressions in separate loops

tcmoore3 committed:
Rename sdf to sdf_compression in python API

tcmoore3 committed:
Merge remote-tracking branch 'origin/trunk-minor' into feature/expansive-sdf

tcmoore3 committed:
Deprecate SDF.sdf for SDF.sdf_compression

tcmoore3 committed:
Document how pressure is related to both SDFs

tcmoore3 committed:
Fix SDF docs for expansive perturbations

tcmoore3 committed:
Remove warning about SDF for concave particles

tcmoore3 committed:
Remove nan checks in SDF

tcmoore3 committed:
Provide x values used in SDF calculation

tcmoore3 committed:
Fix warning

tcmoore3 committed:
Fix bug

tcmoore3 committed:
Add some references to SDF docs

tcmoore3 committed:
Reword SDF docs for expansive and compressive SDFs

tcmoore3 committed:
Merge remote-tracking branch 'origin/trunk-minor' into feature/expansive-sdf

tcmoore3 committed:
Add sdf logger tests

tcmoore3 committed:
Fix mismatching array length error

tcmoore3 committed:
Fix sphinx doc build error

tcmoore3 committed:
Merge branch 'feature/expansive-sdf' of github.com:glotzerlab/hoomd-blue into feature/expansive-sdf

joaander commented on hoomd/hpmc/compute.py:
```diff
@@ -295,28 +348,80 @@ def _attach_hook(self):
+        distribution function for compression moves \
+        :math:`[\\mathrm{probability\\ density}]`.
+
+        .. deprecated:: v3.8.0
```
```suggestion
        .. deprecated:: v3.11.0
```
The next feature release will be v3.11.0.

joaander commented on hoomd/hpmc/compute.py:
```diff
@@ -295,28 +348,80 @@ def _attach_hook(self):
+        .. deprecated:: v3.8.0
+            Use `sdf_compression`.
+        """
+        self._cpp_obj.compute(self._simulation.timestep)
+        return self._cpp_obj.sdf_compression
```
```suggestion
        return self.sdf_compression
```
Reuse existing code.

joaander commented on hoomd/hpmc/compute.py:
```diff
@@ -325,18 +430,37 @@ def betaP(self):  # noqa: N802 - allow function name
             `betaP` is `None` on ranks >= 1.
         """
-        if not numpy.isnan(self.sdf).all():
+        if not (numpy.isnan(self.sdf_compression).all()
+                or numpy.isnan(self.sdf_expansion).all()):
```
```suggestion
        if self.sdf_compression is not None and self.sdf_expansion is not None:
```
Better yet,
```
sdf_compression = self.sdf_compression;
sdf_expansion = self.sdf_expansion;
if self.sdf_compression is not None and self.sdf_expansion is not None:
```
and then use `sdf_*` below instead of `self.sdf_*` to avoid making unnecessary data copies.

joaander commented on hoomd/hpmc/compute.py:
```diff
@@ -325,18 +430,37 @@ def betaP(self):  # noqa: N802 - allow function name
-            sdf_fit = numpy.array(self.sdf)
-            # construct the x coordinates
-            x_fit = numpy.arange(0, len(sdf_fit), 1) * self.dx
+            sdf_fit_compression = numpy.array(self.sdf_compression)
+            x_fit = numpy.arange(0, len(sdf_fit_compression), 1) * self.dx
             x_fit += self.dx / 2
```
Reuse code, use `self.x_compression`.

joaander commented on hoomd/hpmc/compute.py:
```diff
@@ -325,18 +430,37 @@ def betaP(self):  # noqa: N802 - allow function name
+
+            # expansive contribution
+            # reverse the x-coordinates
+            x_fit = -x_fit[::-1]
```
This assumes that `len(sdf_compression) == len(sdf_expansion)`. Perhaps add an assert here to be certain.

joaander commented on hoomd/hpmc/compute.py:
```diff
@@ -295,28 +348,80 @@ def _attach_hook(self):
+        distribution function for the compressive perturbations \
+        :math:`[\\mathrm{length}]`."""
+        sdf_compression = numpy.array(self.sdf_compression)
+        x = numpy.arange(0, len(sdf_compression), 1) * self.dx
```
If it doesn't already exist, add a new C++ method to get the number of bins. Call that instead of `len(self.sdf_compression)`. This avoids making numerous extra copies of the histogram (see below).

joaander commented on hoomd/hpmc/pytest/test_compute_sdf.py:
```diff
@@ -286,5 +286,13 @@ def test_logging():
+            'sdf_expansion': {
+                'category': LoggerCategories.sequence,
+                'default': True
+            },
```
Add `x_compression` and `x_expansion` to this list.

joaander commented on hoomd/hpmc/pytest/test_compute_sdf.py:
```diff
@@ -270,7 +270,7 @@ def test_linear_search_path(simulation_factory, two_particle_snapshot_factory):
     # that the SDF is zero everywhere else
```
Add unit tests for `sdf_expansion`.

joaander reviewed (changes requested):
Thanks! This is looking a lot better. I look forward to seeing the validation test results and a new test workflow in https://github.com/glotzerlab/hoomd-validation/. There are some small changes needed to polish this code off (presuming the validation tests pass).

joaander committed:
Merge branch 'trunk-minor' into feature/expansive-sdf

joaander committed:
Revise documentation.

joaander committed:
Compute histograms only once.

Provide consistent results between sdf_expansion and x_expansion.

joaander committed:
Correct definition of k for x_expansion.

joaander committed:
Additional unit tests and fixes.

joaander committed:
Run pre-commit.

tcmoore3 commented on hoomd/hpmc/compute.py:
```diff
@@ -182,11 +202,31 @@ class SDF(Compute):
+            \} &
+
+
+    For particle :math:`i`, `SDF` finds th the minimum (maximum for expansive
```
```suggestion
    For particle :math:`i`, `SDF` finds the the minimum (maximum for expansive
```

tcmoore3 committed:
Update hoomd/hpmc/compute.py

joaander committed:
Return None on non-root ranks as documented.

joaander reviewed (approved):
Thanks!

joaander commented:
Validation tests will be provided in separate work on `hoomd-validation`: https://github.com/glotzerlab/hoomd-validation/issues/43
